Add normalisation

leilafarsani opened the pull request:
Add percentage normalization for countries and sectors

This PR implements percentage normalization for both country and sector splits. The changes include:

Changes made:
- Implemented country percentage normalisation in `_get_recipient_countries_with_normalised_percentages`
- Implemented sector percentage normalisation in `_get_sectors_grouped_by_vocab_with_normalised_percentages`
- Added test `test_no_double_counting` to verify:
  - Total values match the original amount
  - Country percentages are correctly normalised
  - Sector percentages are correctly normalised
  - No double counting occurs

Test Results:
All tests passed successfully:

TODO:
- Region splitting functionality will be implemented in a separate PR

odscjames commented on iati_activity_details_split_by_fields/iati_activity.py:
```diff
@@ -72,17 +72,43 @@ def get_transactions_split_as_json(self):
+        if not self.recipient_countries:
+            return []
+        total_percentage = sum(
+            country.percentage or 0 
```
Pedantically this is vulnerable to people doing minus values, could also do `max(country.percentage, 0)` (I think, not tested) but I think at this stage we'll ignore data that that's wrong until we see signs it's an actual problem.

odscjames commented on tests/test_at_activity_level.py:
```diff
@@ -186,3 +186,44 @@ def test_split_by_everything():
     # It's possible to verify this by hand,
     # but may as well get Python to check for us and avoid extra work and the possibility of mistakes
     # (Can use in other tests too)
```
Can add a line here to say this is done in the test_no_double_counting test

odscjames commented:
Nice, thank you! Good catch on using copy.deepcopy before changing data in place.

Just some minor requests:
* Lint errors to fix;  see failed lint action and listed in https://github.com/IATI/activity-details-split-by-fields/blob/main/.github/workflows/lint.yml too
* Can we have 2 tests to check the % normalisation? Basically `test_split_by_country` and `test_split_by_sector` but with dodgy % data as the inputs. Could copy tests then edit or set up `@pytest.mark.parametrize` to avoid copying to much code. https://docs.pytest.org/en/stable/how-to/parametrize.html#pytest-mark-parametrize-parametrizing-test-functions

leilafarsani commented:
Thank you @odscjames - updates done:

- Fixed lint errors
- Added two other tests for normalising incorrect percentages (both country and sector)
- Added note about double counting test

All tests are now passing.
